File: tests/parity/regen_golden.py
"""regen_golden.py - Raw and normalized pycone pipeline for golden fixtures.

Reads a JSON payload describing a resolved set of pycone inputs from
stdin and writes the four pipeline stages (absorbance, absorptance,
corneal quantal, corneal energy) as CSV to stdout. Used by
tests/parity/regen_golden.m to source the golden fixtures in
tests/data/ from the vendored pycone reference.

This script computes the stages WITHOUT pycone's internal per-stage
peak renormalization, so it can emit the raw (un-normalized) values the
toolbox produces when NormalizeOutput is false. When a normalization
grid is supplied, each stage is divided by its peak over that grid,
matching the toolbox "Sampled" normalization. All template, macular,
lens and self-screening formulas are taken from the vendored pycone
modules (CMFtemplates, CMFcalc), so the fixtures remain an independent
reference for the toolbox.

Input JSON schema (one object):
    {
      "Lshift":       float,
      "Mshift":       float,
      "Sshift":       float,
      "Lod":          float,
      "Mod":          float,
      "Sod":          float,
      "mac_density":  float,
      "lens_density": float,
      "L_template":   "Lmean" | "Lser" | "Lala" | "M-in-L",
      "M_template":   "Standard" | "L-in-M",
      "wl_min":       float,
      "wl_max":       float,
      "wl_step":      float,
      "norm_min":     float (optional),
      "norm_max":     float (optional),
      "norm_step":    float (optional)
    }

When norm_min/norm_max/norm_step are present the absorptance, quantal
and energy stages are normalized to their peak over that grid; the
absorbance stage is never normalized (the template is already unit at
its true lambda_max). Output rows cover wl_min:wl_step:wl_max.

Output CSV columns:
    nm,
    L_absorbance, M_absorbance, S_absorbance,
    L_absorptance, M_absorptance, S_absorptance,
    L_quantal,    M_quantal,    S_quantal,
    L_energy,     M_energy,     S_energy
"""

import json
import os
import sys

import numpy as np

HERE = os.path.dirname(os.path.abspath(__file__))
PYCONE_DIR = os.path.join(HERE, "pycone")
sys.path.insert(0, PYCONE_DIR)

import CMFtemplates  # noqa: E402

LSER_M_LMAX_DIFF = 23.67
MAC_TEMPLATE_460 = 0.35
LENS_TEMPLATE_400 = 1.7649

LN10 = np.log(10.0)


def pow10(x):
    # Evaluate 10**x as exp(x*log(10)) to match MATLAB's 10.^x libm
    # convention. The two differ by about one unit in the last place,
    # which is amplified by the catastrophic cancellation in
    # 1 - 10**(-od*A) for the deep spectral tails; matching the
    # convention keeps the stages numerically aligned with the toolbox
    # while leaving every spectral input sourced from pycone.
    return np.exp(np.asarray(x) * LN10)


def template_lin(nm, cfg):
    l_template = cfg["L_template"]
    lshift = cfg["Lshift"]
    mshift = cfg["Mshift"]
    sshift = cfg["Sshift"]

    if l_template == "Lser":
        l_log = CMFtemplates.Lserconelog(nm, lshift)
    elif l_template == "Lala":
        l_log = CMFtemplates.Lserconelog(nm, lshift - 2.7)
    elif l_template == "Lmean":
        if lshift == 0:
            l_log = CMFtemplates.Lmeanconelog(nm)
        else:
            lser = CMFtemplates.Lserconelog(nm, lshift)
            lala = CMFtemplates.Lserconelog(nm, lshift - 2.7)
            l_log = np.log10(0.56 * pow10(lser) + 0.44 * pow10(lala))
    elif l_template == "M-in-L":
        l_log = CMFtemplates.Mconelog(nm, lshift + LSER_M_LMAX_DIFF)
    else:
        raise ValueError("unknown L template: " + repr(l_template))

    m_template = cfg.get("M_template", "Standard")
    if m_template == "Standard":
        m_log = CMFtemplates.Mconelog(nm, mshift)
    elif m_template == "L-in-M":
        m_log = CMFtemplates.Lserconelog(nm, mshift - LSER_M_LMAX_DIFF)
    else:
        raise ValueError("unknown M template: " + repr(m_template))

    s_log = CMFtemplates.Sconelog(nm, sshift)
    return np.column_stack([pow10(l_log), pow10(m_log), pow10(s_log)])


def raw_stages(nm, cfg):
    absorbance = template_lin(nm, cfg)

    od = np.array([cfg["Lod"], cfg["Mod"], cfg["Sod"]])
    absorptance = (1.0 - pow10(-od[None, :] * absorbance)) / (1.0 - pow10(-od))[None, :]

    macs = CMFtemplates.macular(nm)
    lenss = CMFtemplates.lens(nm)
    macscale = cfg["mac_density"] / MAC_TEMPLATE_460
    lensscale = cfg["lens_density"] / LENS_TEMPLATE_400
    transmission = pow10((macs * macscale)[:, None]) * pow10((lenss * lensscale)[:, None])
    quantal = absorptance / transmission

    energy = quantal * nm[:, None]

    return absorbance, absorptance, quantal, energy


def main():
    cfg = json.loads(sys.stdin.read())

    nm_min, nm_max, nm_step = cfg["wl_min"], cfg["wl_max"], cfg["wl_step"]
    n_points = int(round((nm_max - nm_min) / nm_step)) + 1
    nm = np.linspace(nm_min, nm_max, n_points)

    absorbance, absorptance, quantal, energy = raw_stages(nm, cfg)

    if "norm_min" in cfg:
        n_norm = int(round((cfg["norm_max"] - cfg["norm_min"]) / cfg["norm_step"])) + 1
        nm_norm = np.linspace(cfg["norm_min"], cfg["norm_max"], n_norm)
        _, norm_absorptance, norm_quantal, norm_energy = raw_stages(nm_norm, cfg)
        absorptance = absorptance / np.max(norm_absorptance, axis=0)[None, :]
        quantal = quantal / np.max(norm_quantal, axis=0)[None, :]
        energy = energy / np.max(norm_energy, axis=0)[None, :]

    out = np.column_stack([nm, absorbance, absorptance, quantal, energy])
    np.savetxt(
        sys.stdout, out, delimiter=",",
        header=("nm,"
                "L_absorbance,M_absorbance,S_absorbance,"
                "L_absorptance,M_absorptance,S_absorptance,"
                "L_quantal,M_quantal,S_quantal,"
                "L_energy,M_energy,S_energy"),
        comments="", fmt="%.17g",
    )


if __name__ == "__main__":
    main()

File: tests/parity/regen_golden.m
function regen_golden()
% REGEN_GOLDEN  Regenerate golden fixtures in tests/data from pycone.
%
%   Sources every regenerated spectral value from the vendored pycone
%   reference (tests/parity/pycone) via tests/parity/regen_golden.py, so
%   the fixtures stay an independent cross-check of the toolbox rather
%   than a copy of its own output. For each fixture this driver
%   constructs the same IndividualCMF observer the consuming test uses,
%   reads the resolved biophysical values (lambda-max shifts, optical
%   densities, macular and lens densities, opsin templates), and passes
%   those exact numbers to pycone. Pure-input columns (for example the
%   genotype-driven shifts) are written from the resolved observer.
%
%   Run this whenever the vendored pycone reference moves and the stored
%   golden CSVs need to be refreshed.
%
%   SPDX-License-Identifier: AGPL-3.0-or-later
%   Copyright (c) 2025-2026 Alexander Forsythe and Brian Funt

here = fileparts(mfilename('fullpath'));
dataDir = fullfile(here, '..', 'data');
pyScript = fullfile(here, 'regen_golden.py');

if ~isfolder(fullfile(here, 'pycone'))
    error('regen_golden:NoPycone', ...
        'pycone reference not found at %s', fullfile(here, 'pycone'));
end

addpath(fullfile(here, '..', '..', 'toolbox'));

warnIDs = { ...
    'StockmanRiderPhotopigmentTemplate:ShiftOverride', ...
    'IndividualCMF:WavelengthOutOfRange', ...
    'Nomograms:WavelengthOutOfRange'};
for k = 1:numel(warnIDs)
    warning('off', warnIDs{k});
end
cleanup = onCleanup(@() reenableWarnings(warnIDs)); %#ok<NASGU>

writeStandard(pyScript, dataDir, 2, 'standard_2deg.csv');
writeStandard(pyScript, dataDir, 10, 'standard_10deg.csv');
writeTemplate(pyScript, dataDir);
writeIndividual(pyScript, dataDir);
writePipeline(pyScript, dataDir);
writeRGB(pyScript, dataDir);
writeRGBMatrix(pyScript, dataDir);
writeCmfSColumn(pyScript, dataDir);

fprintf('Golden fixtures regenerated from pycone.\n');

end


function writeStandard(pyScript, dataDir, observer, fileName)
% nm,L_energy,M_energy,S_energy,L_quantal,M_quantal,S_quantal on 390:5:780.
obs = IndividualCMF(StandardObserver=observer);
py = runPycone(pyScript, payloadFor(obs, 390, 780, 5));

T = table();
T.nm = py.nm;
T.L_energy = py.L_energy;
T.M_energy = py.M_energy;
T.S_energy = py.S_energy;
T.L_quantal = py.L_quantal;
T.M_quantal = py.M_quantal;
T.S_quantal = py.S_quantal;
writeTableExact(T, fullfile(dataDir, fileName));
end


function writeTemplate(pyScript, dataDir)
% wavelength,l_mean_absorbance,l_serine_absorbance,l_hybrid_absorbance
% on 390:1:780. Each column is the L-cone absorbance stage for a
% different L opsin template.
obs = IndividualCMF(StandardObserver=2);

obs.L_OpsinTemplate = "Mean";
pyMean = runPycone(pyScript, payloadFor(obs, 390, 780, 1));
obs.L_OpsinTemplate = "Serine";
pySer = runPycone(pyScript, payloadFor(obs, 390, 780, 1));
obs.L_OpsinTemplate = "MinL";
pyHyb = runPycone(pyScript, payloadFor(obs, 390, 780, 1));

T = table();
T.wavelength = pyMean.nm;
T.l_mean_absorbance = pyMean.L_absorbance;
T.l_serine_absorbance = pySer.L_absorbance;
T.l_hybrid_absorbance = pyHyb.L_absorbance;
writeTableExact(T, fullfile(dataDir, 'template_verification.csv'));
end


function writeIndividual(pyScript, dataDir)
% nm,L_energy,M_energy,S_energy,L_quantal,M_quantal,S_quantal on
% 390:5:780, normalized over the 380:5:780 sampling grid used by the
% consuming test's Sampled normalization.
obs = IndividualCMF( ...
    'L_OpsinTemplate', "Serine", ...
    'L_LambdaMaxShift', 5.0, ...
    'M_LambdaMaxShift', -5.0, ...
    'Lod', 0.60, ...
    'Mod', 0.60, ...
    'Sod', 0.45, ...
    'MacularDensity', 0.50, ...
    'LensDensity', 2.00);

payload = payloadFor(obs, 390, 780, 5);
payload.norm_min = 380;
payload.norm_max = 780;
payload.norm_step = 5;
py = runPycone(pyScript, payload);

T = table();
T.nm = py.nm;
T.L_energy = py.L_energy;
T.M_energy = py.M_energy;
T.S_energy = py.S_energy;
T.L_quantal = py.L_quantal;
T.M_quantal = py.M_quantal;
T.S_quantal = py.S_quantal;
writeTableExact(T, fullfile(dataDir, 'individual_verification.csv'));
end


function writePipeline(pyScript, dataDir)
% nm,L_log_abs,M_log_abs,L_absorptance,M_absorptance,
% L_quantal,M_quantal,S_quantal,L_hybrid_energy,M_hybrid_energy
% on 390:5:780. Standard 2-deg observer for the raw stages, plus two
% hybrid observers for the swapped-template energy columns.
obs = IndividualCMF(StandardObserver=2);
py = runPycone(pyScript, payloadFor(obs, 390, 780, 5));

obsL = IndividualCMF(Age=32, FieldSize=2);
obsL.L_OpsinTemplate = "MinL";
pyL = runPycone(pyScript, payloadFor(obsL, 390, 780, 5));

obsM = IndividualCMF(Age=32, FieldSize=2);
obsM.M_OpsinTemplate = "LinM";
pyM = runPycone(pyScript, payloadFor(obsM, 390, 780, 5));

T = table();
T.nm = py.nm;
T.L_log_abs = log10(py.L_absorbance);
T.M_log_abs = log10(py.M_absorbance);
T.L_absorptance = py.L_absorptance;
T.M_absorptance = py.M_absorptance;
T.L_quantal = py.L_quantal;
T.M_quantal = py.M_quantal;
T.S_quantal = py.S_quantal;
T.L_hybrid_energy = pyL.L_energy;
T.M_hybrid_energy = pyM.M_energy;
writeTableExact(T, fullfile(dataDir, 'pipeline_verification.csv'));
end


function writeRGB(pyScript, dataDir)
% nm,R,G,B,L_quantal,M_quantal,S_quantal on 390:5:780. The RGB columns
% are the normalized-energy color matching functions; the quantal
% columns are the raw corneal quantal sensitivities.
obs = IndividualCMF(StandardObserver=2);
py = runPycone(pyScript, payloadFor(obs, 390, 780, 5));

primaries = obs.Primaries(:);
primEnergy = pointEnergy(pyScript, obs, primaries);

% Per-cone energy peaks over the spectrum grid normalize the primaries
% consistently with the normalized LMS energy used for the CMFs.
peaks = [max(py.L_energy), max(py.M_energy), max(py.S_energy)];
primNorm = primEnergy ./ peaks;
lmsNorm = [py.L_energy, py.M_energy, py.S_energy] ./ peaks;

% Rows of primNorm are primaries, columns are LMS.
rgbCmfs = (primNorm.' \ lmsNorm.').';

T = table();
T.nm = py.nm;
T.R = rgbCmfs(:, 1);
T.G = rgbCmfs(:, 2);
T.B = rgbCmfs(:, 3);
T.L_quantal = py.L_quantal;
T.M_quantal = py.M_quantal;
T.S_quantal = py.S_quantal;
writeTableExact(T, fullfile(dataDir, 'rgb_verification.csv'));
end


function writeRGBMatrix(pyScript, dataDir)
% Headerless 3x3 matrix inv([L;M;S]) built from the raw corneal energy
% at the three primary wavelengths (Standard 2-deg observer).
obs = IndividualCMF(StandardObserver=2);
primaries = obs.Primaries(:);
primEnergy = pointEnergy(pyScript, obs, primaries);

% primEnergy rows are primaries, columns are LMS. The consuming test
% forms [L_prim; M_prim; S_prim] (rows LMS, columns primaries) and
% inverts it.
RGBLMS = primEnergy.';
M = inv(RGBLMS);
writeMatrixExact(M, fullfile(dataDir, 'rgb_matrix.csv'));
end


function writeMatrixExact(M, path)
% Write a numeric matrix at full double precision, comma-separated.
fid = fopen(path, 'w');
cleanup = onCleanup(@() fclose(fid)); %#ok<NASGU>
for r = 1:size(M, 1)
    fprintf(fid, '%.18e', M(r, 1));
    for c = 2:size(M, 2)
        fprintf(fid, ',%.18e', M(r, c));
    end
    fprintf(fid, '\n');
end
end


function writeCmfSColumn(pyScript, dataDir)
% Refresh only the stale S_log column of cmf_verification_data.csv (the
% S photopigment template coefficients changed). Every other column is
% preserved byte-for-byte: the original lines are read as text and only
% the S_log field is replaced in place.
csvPath = fullfile(dataDir, 'cmf_verification_data.csv');
lines = readlines(csvPath);
if lines(end) == ""
    lines(end) = [];
end

header = split(lines(1), ',');
sCol = find(header == "S_log", 1);
if isempty(sCol)
    error('regen_golden:NoSLogColumn', 'S_log column not found in %s', csvPath);
end
nmCol = find(header == "nm", 1);

nRows = numel(lines) - 1;
nmValues = zeros(nRows, 1);
for r = 1:nRows
    fields = split(lines(r + 1), ',');
    nmValues(r) = str2double(fields(nmCol));
end
nmMin = min(nmValues);
nmMax = max(nmValues);
nmStep = nmValues(2) - nmValues(1);

obs = IndividualCMF(StandardObserver=2);
py = runPycone(pyScript, payloadFor(obs, nmMin, nmMax, nmStep));
sLog = log10(py.S_absorbance);

for r = 1:nRows
    fields = split(lines(r + 1), ',');
    fields(sCol) = string(sprintf('%.17g', sLog(r)));
    lines(r + 1) = strjoin(fields, ',');
end
writelines(lines, csvPath);
end


function E = pointEnergy(pyScript, obs, wavelengths)
% Raw corneal energy at arbitrary wavelengths, one pycone call per
% point so non-uniform primary grids are handled exactly. Returns an
% N-by-3 matrix with rows matching the input wavelengths and columns
% L, M, S.
E = zeros(numel(wavelengths), 3);
for i = 1:numel(wavelengths)
    payload = payloadFor(obs, wavelengths(i), wavelengths(i), 1);
    py = runPycone(pyScript, payload);
    E(i, :) = [py.L_energy(1), py.M_energy(1), py.S_energy(1)];
end
end


function payload = payloadFor(obs, wlMin, wlMax, wlStep)
payload = struct( ...
    'Lshift', obs.L_LambdaMaxShift, ...
    'Mshift', obs.M_LambdaMaxShift, ...
    'Sshift', obs.S_LambdaMaxShift, ...
    'Lod', obs.Lod, ...
    'Mod', obs.Mod, ...
    'Sod', obs.Sod, ...
    'mac_density', obs.MacularDensity, ...
    'lens_density', obs.LensDensity, ...
    'L_template', mapLTemplate(obs.L_OpsinTemplate), ...
    'M_template', mapMTemplate(obs.M_OpsinTemplate), ...
    'wl_min', wlMin, ...
    'wl_max', wlMax, ...
    'wl_step', wlStep);
end


function s = mapLTemplate(matlabName)
switch string(matlabName)
    case "Mean",    s = "Lmean";
    case "Serine",  s = "Lser";
    case "Alanine", s = "Lala";
    case "MinL",    s = "M-in-L";
    otherwise,      s = "Lmean";
end
end


function s = mapMTemplate(matlabName)
switch string(matlabName)
    case "Standard", s = "Standard";
    case "LinM",     s = "L-in-M";
    case "Mean",     s = "Standard";
    otherwise,       s = "Standard";
end
end


function tbl = runPycone(scriptPath, payload)
payloadJson = jsonencode(payload);
tmp = tempname;
writelines(payloadJson, tmp);
cleanup = onCleanup(@() delete(tmp)); %#ok<NASGU>
[status, out] = system(sprintf('python3 %s < %s', scriptPath, tmp));
if status ~= 0
    error('regen_golden:PyconeFailure', ...
        'pycone subprocess failed (status %d):\n%s', status, out);
end
csvTmp = tempname + ".csv";
fid = fopen(csvTmp, 'w');
fprintf(fid, '%s', out);
fclose(fid);
cleanupCsv = onCleanup(@() delete(csvTmp)); %#ok<NASGU>
tbl = readtable(csvTmp);
end


function writeTableExact(T, path)
% Write a table with 17 significant digits so reloaded values round-trip
% to machine precision (writetable's default 15 digits loses relative
% precision on small spectral values).
names = T.Properties.VariableNames;
nCols = numel(names);
nRows = height(T);
data = cell(nRows, nCols);
for c = 1:nCols
    col = T.(names{c});
    for r = 1:nRows
        data{r, c} = sprintf('%.17g', col(r));
    end
end
fid = fopen(path, 'w');
cleanup = onCleanup(@() fclose(fid)); %#ok<NASGU>
fprintf(fid, '%s', strjoin(names, ','));
fprintf(fid, '\n');
for r = 1:nRows
    fprintf(fid, '%s', strjoin(data(r, :), ','));
    fprintf(fid, '\n');
end
end


function reenableWarnings(ids)
for i = 1:numel(ids)
    warning('on', ids{i});
end
end
